### PR TITLE
BuildingSync Measure-Gem

### PR DESCRIPTION
**Pull request by ThibaultMarzullo**
This PR is to allow the BuildingSync Gem to be used as a measure-gem.

This PR solves [Issue #130](https://github.com/BuildingSync/project-tracker/issues/130) in the project tracker.

### CODE REVIEW

**Review by nllong — COMMENTED** [review]
a lot to review here, do you need this done today?

**Comment by ThibaultMarzullo** [comment]
No I don’t think so, but I’ll let Lauren chime in. Thanks!

**Comment by laurenliz22** [comment]
> No I don’t think so, but I’ll let Lauren chime in. Thanks!
> […](#)
> On Sep 30, 2022, at 11:26 AM, Nicholas Long ***@***.***> wrote: ﻿ @nllong commented on this pull request. a lot to review here, do you need this done today? — Reply to this email directly, view it on GitHub, or unsubscribe. You are receiving this because you were assigned.

No - not today.  In the next week or so would be the goal.

File: lib/buildingsync/makers/phase_zero_base.osw [hunk 1]
```diff
@@ -86,36 +66,11 @@
...
-			"arguments": {
-				"__SKIP__": true
-			}
 		}, {
 			"measure_dir_name": "AdjustThermostatSetpointsByDegrees",
 			"arguments": {
 				"__SKIP__": true
 			}
-		}, {
-			"measure_dir_name": "ReplaceFanTotalEfficiency",
```
Comment (nllong) [inline lib/buildingsync/makers/phase_zero_base.osw:110]: For all of these measures that have been removed, did we lose functionality? Or has the functionality just moved?

File: lib/measures/building_sync_to_openstudio/README.md [hunk 2]
```diff
@@ -0,0 +1,46 @@
+# BuildingSync Gem
+
+## Description
+
+This measure-gem converts a BuildingSync XML file into a series of OSWs. Each OSW corresponds to an energy efficiency package of measures defined in the BuildingSync XML file. The OSWs can then be simulated and the results are written back into the BuildingSync XML file.
```
Comment (nllong) [inline lib/measures/building_sync_to_openstudio/README.md:5]: this is great!

File: lib/measures/building_sync_to_openstudio/README.md [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+## Modeler Description
+
+The measure will use a BuildingSync XML file as an input. The XML can be created using tools such as bsyncpy.
```
Comment (nllong) [inline lib/measures/building_sync_to_openstudio/README.md]:
```suggestion
The measure will use a BuildingSync XML file as an input. The XML can be created using tools such as [bsyncpy](https://pypi.org/project/bsync/).
```

File: lib/measures/building_sync_to_openstudio/README.md [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+The XML will be parsed and a new OpenStudio model will be created. A new OSW will be created for each energy efficiency package of measures defined in the XML, using the measures defined in the ./lib/buildingsync/makers/phase_zero_base.osw file.
+The user has a choice for just generating OSWs, or generating and simulating them. If the OSWs are simulated, then the results are collected and reports are generated. These reports will be written in the original BuildingSync XML file, and that XML file will be saved.
```
Comment (nllong) [inline lib/measures/building_sync_to_openstudio/README.md]:
```suggestion
The user has a choice for just generating OSWs or generating and simulating them. If the OSWs are simulated, then the results are collected and reports are generated. These reports will be written in the original BuildingSync XML file, and that XML file will be saved.
```

File: lib/measures/building_sync_to_openstudio/measure.rb [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+  end
+
+  # human readable description of modeling approach
+  def modeler_description
+    return "BuildingSync to OSM translation used to happen outside of the measure structure, as a result it couldn't easily be used in tools that support running OSW's through the OpenStudio CLI. When upgrading to support OpenStudio 3.4, this code was wrapped into a measure. Additionally, where libraries exist in the OpenStudio Extension Gem and OOpenStudio Standars Gem, those should be used vs. custom code within BuildingSync. This will provide consistency with other workflows and minimize upgrade maintenance. This measure doesn't work with an off the shelf OpenStudio install because it requires additional gems. As a result it isn't currently on the Building Component Library (BCL). It's also possible that at some point this measure may need to run other OpenStudio measures. If that happens it does result in extra planning on setting up a project to assure that the necessary measures are available, possibly through bundle and gem files for projects using this."
```
Comment (nllong) [inline lib/measures/building_sync_to_openstudio/measure.rb]: I don't think we need all of the 'history' in this description. Just say what this measure does now.

Comment (nllong) [reply lib/measures/building_sync_to_openstudio/measure.rb]: After making these changes, you will need to rerun the "update_measures" method to regenerate the readme and measure.xml file.

File: lib/measures/building_sync_to_openstudio/measure.rb [hunk 6]
```diff
@@ -0,0 +1,114 @@
...
+    # path/to/output_dir/scenario_name
+    translator.write_osws
+    if simulate_flag
+      # run all simulations
+      translator.run_osws
+
+      # gather the results for all scenarios found in out_path,
+      # such as annual and monthly data for different energy
+      # sources (electricity, natural gas, etc.)
+      translator.gather_results()
```
Comment (nllong) [inline lib/measures/building_sync_to_openstudio/measure.rb]:
```suggestion
      translator.gather_results
```

**Review by nllong — APPROVED** [review]
approving but there are a few tweaks that would be nice.

**Comment by nllong** [comment]
Any reason not to merge this?

**Pull request merged**